mtsp/src/main.leo

```diff
@@ -0,0 +1,792 @@
+      authorized_until: input_record.authorized_until
+    };
+
+    let external_authorization_required: bool = input_record.external_authorization_required;
+    let authorized_until: u32 = external_authorization_required ? 0u32 : 4294967295u32;
+    let transfer_record: Token = Token {
+      owner: recipient,
+      amount: amount,
+      token_id: input_record.token_id,
+      external_authorization_required: external_authorization_required,
+      authorized_until: authorized_until
+    };
+
+    return (updated_record, transfer_record, finalize_transfer_private(external_authorization_required, input_record.authorized_until));
+  }
+
+  async function finalize_transfer_private(
+    external_authorization_required: bool,
+    input_token_authorized_until: u32
+  ) {
+    assert(block.height <= input_token_authorized_until || !external_authorization_required);
+  }
+
+  async transition transfer_private_to_public(
+    recipient: address,
+    amount: u128,
+    input_record: Token
+  ) -> (Token, Future) {
+    let updated_record: Token = Token {
+      owner: input_record.owner,
+      amount: input_record.amount - amount,
+      token_id: input_record.token_id,
+      external_authorization_required: input_record.external_authorization_required,
+      authorized_until: input_record.authorized_until
+    };
+
+    return (updated_record, finalize_transfer_private_to_public(input_record.token_id, recipient, amount));
+  }
+
+  async function finalize_transfer_private_to_public(token_id: field, recipient: address, amount: u128) {
+    // Get or create the balance for the recipient
+    let recipient_key: TokenOwner = TokenOwner {
+      account: recipient,
+      token_id: token_id
+    };
+    let recipient_balance_key: field = BHP256::hash_to_field(recipient_key);
+    // Get the locked balance if authorization is required, otherwise get the authorized balance
+    let authorization_required: bool = registered_tokens.get(token_id).external_authorization_required;
+    let default_expiration: u32 = authorization_required ? 0u32 : 4294967295u32;
+    let default_balance: Balance = Balance {
+      token_id: token_id,
+      account: recipient,
+      balance: 0u128,
+      authorized_until: default_expiration
+    };
+    let recipient_balance: Balance = authorization_required ? balances.get_or_use(recipient_balance_key, default_balance) : authorized_balances.get_or_use(recipient_balance_key, default_balance);
+    let new_recipient_balance: Balance = Balance {
+      token_id: token_id,
+      account: recipient,
+      balance: recipient_balance.balance + amount,
+      authorized_until: recipient_balance.authorized_until
+    };
+    // Update the balance of the recipient
+    if (authorization_required) {
+      balances.set(recipient_balance_key, new_recipient_balance);
+    } else {
+      authorized_balances.set(recipient_balance_key, new_recipient_balance);
+    }
+  }
+
+  // -------------------------
+  // credits.aleo wrapper
```

Comment:
For this wrapper code, I think we can delete the methods below. Instead we should have a `deposit_credits_public` , `deposit_credits_private`, `withdraw_credits_public`, `withdraw_credits_private`. These methods should enable someone to wrap credits into this program so that a DeFi protocol doesn't need special handling for the `credits.aleo` program which won't be able to ever initialize a token or mint it.